### DISCUSSION TITLE
Broken HTTP2 cancellations tests

### DISCUSSION THREAD

**Discussion by karpetrosyan** [encode/httpcore#755] [Potential Issue]
Sometimes http2 connection cancellation tests fail, but other times they succeed.

The issue appears to be in `AsyncHTTP2Connection's` `async_handle_request` method, which cannot always handle cancelled errors.

If the cancellation happens in [this section of the code](https://github.com/encode/httpcore/blob/e32b2adb21bc38a2ca305f8dac1de153635a2b16/httpcore/_async/http2.py#L114C1-L130), the connection will not call the response_closed method.

**Comment by lovelydinosaur** [comment]
(As a starting point...)

> Sometimes http2 connection cancellation tests fail, but other times they succeed.

Could you provide a link/example for this?

**Reply by zanieb** [reply]
https://github.com/encode/httpcore/actions/runs/5532888849/jobs/10095604921#step:7:66

**Comment by karpetrosyan** [comment]
You can get that error by inserting `anyio.sleep` in the middle of this code block.

Like so

```python

                self._sent_connection_init = True

                # Initially start with just 1 until the remote server provides
                # its max_concurrent_streams value
                self._max_streams = 1

                local_settings_max_streams = (
                    self._h2_state.local_settings.max_concurrent_streams
                )
                self._max_streams_semaphore = AsyncSemaphore(local_settings_max_streams)
                await anyio.sleep(0.5)
                for _ in range(local_settings_max_streams - self._max_streams):
                    await self._max_streams_semaphore.acquire()

        await self._max_streams_semaphore.acquire()

```

This line may cause a Canceled Error, which we do not handle. We use timeout 0.01 in tests, so I frequently get failing tests on my local machine.

``` python
await self._max_streams_semaphore.acquire()
```

**Comment by karpetrosyan** [comment]
Issue:  https://github.com/encode/httpcore/issues/756
